Hardware/CPU/AMD17CPU.cs

```diff
@@ -0,0 +1,476 @@
+        if (core == null)
+          return;        
+        CPUID cpu = core.Threads[0];
+        if (cpu == null)
+          return;
+        uint eax, edx;
+
+        ulong mask = Ring0.ThreadAffinitySet(1UL << cpu.Thread);
+
+        // MSRC001_0299 
+        // TU [19:16] 
+        // ESU [12:8] -> Unit 15.3 micro Joule per increment 
+        // PU [3:0] 
+        Ring0.Rdmsr(MSR_PWR_UNIT, out eax, out edx);
+        int tu = (int)((eax >> 16) & 0xf);
+        int esu = (int)((eax >> 12) & 0xf);
+        int pu = (int)(eax & 0xf);
+
+        // MSRC001_029B 
+        // total_energy [31:0] 
+        DateTime sample_time = DateTime.Now;
+        Ring0.Rdmsr(MSR_PKG_ENERGY_STAT, out eax, out edx);
+        uint total_energy = eax;
+
+        // THM_TCON_CUR_TMP 
+        // CUR_TEMP [31:21] 
+        uint temperature = 0;
+        Ring0.WritePciConfig(Ring0.GetPciAddress(0, 0, 0), FAMILY_17H_PCI_CONTROL_REGISTER, F17H_M01H_THM_TCON_CUR_TMP);
+        Ring0.ReadPciConfig(Ring0.GetPciAddress(0, 0, 0), FAMILY_17H_PCI_CONTROL_REGISTER + 4, out temperature);
+
+        // SVI0_TFN_PLANE0 [0] 
+        // SVI0_TFN_PLANE1 [1] 
+        uint smusvi0_tfn = 0;
+        Ring0.WritePciConfig(Ring0.GetPciAddress(0, 0, 0), FAMILY_17H_PCI_CONTROL_REGISTER, F17H_M01H_SVI + 0x8);
+        Ring0.ReadPciConfig(Ring0.GetPciAddress(0, 0, 0), FAMILY_17H_PCI_CONTROL_REGISTER + 4, out smusvi0_tfn);
+
+        // SVI0_PLANE0_VDDCOR [24:16] 
+        // SVI0_PLANE0_IDDCOR [7:0] 
+        uint smusvi0_tel_plane0 = 0;
+        Ring0.WritePciConfig(Ring0.GetPciAddress(0, 0, 0), FAMILY_17H_PCI_CONTROL_REGISTER, F17H_M01H_SVI + 0xc);
+        Ring0.ReadPciConfig(Ring0.GetPciAddress(0, 0, 0), FAMILY_17H_PCI_CONTROL_REGISTER + 4, out smusvi0_tel_plane0);
+
+        // SVI0_PLANE1_VDDCOR [24:16] 
+        // SVI0_PLANE1_IDDCOR [7:0] 
+        uint smusvi0_tel_plane1 = 0;
+        Ring0.WritePciConfig(Ring0.GetPciAddress(0, 0, 0), FAMILY_17H_PCI_CONTROL_REGISTER, F17H_M01H_SVI + 0x10);
+        Ring0.ReadPciConfig(Ring0.GetPciAddress(0, 0, 0), FAMILY_17H_PCI_CONTROL_REGISTER + 4, out smusvi0_tel_plane1);
+
+        Ring0.ThreadAffinitySet(mask);
+
+        // power consumption 
+        // power.Value = (float) ((double)pu * 0.125); 
+        // esu = 15.3 micro Joule per increment 
+        if (_lastPwrTime.Ticks == 0)
+        {
+          _lastPwrTime = sample_time;
+          _lastPwrValue = total_energy;
+        }
+        // ticks diff 
+        TimeSpan time = sample_time - _lastPwrTime;
+        long pwr;
+        if (_lastPwrValue <= total_energy)
+          pwr = total_energy - _lastPwrValue;
+        else
+          pwr = (0xffffffff - _lastPwrValue) + total_energy;
+
+        // update for next sample 
+        _lastPwrTime = sample_time;
+        _lastPwrValue = total_energy;
+
+        double energy = 15.3e-6 * pwr;
+        energy /= time.TotalSeconds;
+
+        _packagePower.Value = (float)energy;
+        _hw.ActivateSensor(_packagePower);
```

Comment:
While OHM's ListSet (underlying collection for Sensors) doesn't add duplicate sensors, I think this ActivateSensor can be moved to the constructor as it always exists (same with the other 5 sensors in the constructor).